### ISSUE TITLE
[Bug] MMPOSE中configs/body_2d_keypoint/integral_regression/coco/ipr_res50_dsnt-8xb64-210e_coco-256x256.py运行结果为loss: nan  loss_kpt: nan  acc_pose: 0.0000 00

### ISSUE THREAD

**Issue by Limbo844** [open-mmlab/mmpose#2517]
### Prerequisite

- [X] I have searched [Issues](https://github.com/open-mmlab/mmpose/issues) and [Discussions](https://github.com/open-mmlab/mmpose/discussions) but cannot get the expected help.
- [X] The bug has not been fixed in the latest version(https://github.com/open-mmlab/mmpose).

### Environment

运行环境： 3060TI python3.8 MMpose
train_dataloader： batch_size=32 ，其余均为最新源代码

### Reproduces the problem - code sample

_base_ = ['../../../_base_/default_runtime.py']

# runtime
train_cfg = dict(max_epochs=210, val_interval=10)

# optimizer
optim_wrapper = dict(optimizer=dict(
    type='Adam',
    lr=5e-5,
))

# learning policy
param_scheduler = [
    dict(
        type='LinearLR', begin=0, end=500, start_factor=0.001,
        by_epoch=False),  # warm-up
    dict(
        type='MultiStepLR',
        begin=0,
        end=train_cfg['max_epochs'],
        milestones=[170, 200],
        gamma=0.1,
        by_epoch=True)
]

# automatically scaling LR based on the actual training batch size
auto_scale_lr = dict(base_batch_size=512)

# codec settings
codec = dict(
    type='IntegralRegressionLabel',
    input_size=(256, 256),
    heatmap_size=(64, 64),
    sigma=2.0,
    normalize=True)

# model settings
model = dict(
    type='TopdownPoseEstimator',
    data_preprocessor=dict(
        type='PoseDataPreprocessor',
        mean=[123.675, 116.28, 103.53],
        std=[58.395, 57.12, 57.375],
        bgr_to_rgb=True),
    backbone=dict(
        type='ResNet',
        depth=50,
    ),
    head=dict(
        type='DSNTHead',
        in_channels=2048,
        in_featuremap_size=(8, 8),
        num_joints=17,
        loss=dict(
            type='MultipleLossWrapper',
            losses=[
                dict(type='SmoothL1Loss', use_target_weight=True),
                dict(type='JSDiscretLoss', use_target_weight=True)
            ]),
        decoder=codec),
    test_cfg=dict(
        flip_test=True,
        shift_coords=True,
        shift_heatmap=True,
    ),
    init_cfg=dict(
        type='Pretrained',
        checkpoint='https://download.openmmlab.com/mmpose/'
        'pretrain_models/td-hm_res50_8xb64-210e_coco-256x192.pth'))

# base dataset settings
dataset_type = 'CocoDataset'
data_mode = 'topdown'
data_root = 'data/coco/'

# pipelines
train_pipeline = [
    dict(type='LoadImage'),
    dict(type='GetBBoxCenterScale'),
    dict(type='RandomFlip', direction='horizontal'),
    dict(type='RandomHalfBody'),
    dict(type='RandomBBoxTransform'),
    dict(type='TopdownAffine', input_size=codec['input_size']),
    dict(type='GenerateTarget', encoder=codec),
    dict(type='PackPoseInputs')
]
test_pipeline = [
    dict(type='LoadImage'),
    dict(type='GetBBoxCenterScale'),
    dict(type='TopdownAffine', input_size=codec['input_size']),
    dict(type='PackPoseInputs')
]

# data loaders
train_dataloader = dict(
    batch_size=32,
    num_workers=2,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=True),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        data_mode=data_mode,
        ann_file='annotations/person_keypoints_train2017.json',
        data_prefix=dict(img='train2017/'),
        pipeline=train_pipeline,
    ))
val_dataloader = dict(
    batch_size=32,
    num_workers=2,
    persistent_workers=True,
    drop_last=False,
    sampler=dict(type='DefaultSampler', shuffle=False, round_up=False),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        data_mode=data_mode,
        ann_file='annotations/person_keypoints_val2017.json',
        bbox_file=f'{data_root}person_detection_results/'
        'COCO_val2017_detections_AP_H_56_person.json',
        data_prefix=dict(img='val2017/'),
        test_mode=True,
        pipeline=test_pipeline,
    ))
test_dataloader = val_dataloader

# hooks
default_hooks = dict(checkpoint=dict(save_best='coco/AP', rule='greater'))

# evaluators
val_evaluator = dict(
    type='CocoMetric',
    ann_file=f'{data_root}annotations/person_keypoints_val2017.json')
test_evaluator = val_evaluator

### Reproduces the problem - command or script

python tools/train.py configs/body_2d_keypoint/integral_regression/coco/ipr_res50_dsnt-8xb64-210e_coco-256x256.py

### Reproduces the problem - error message

07/05 15:23:26 - mmengine - INFO - Epoch(train)   [1][  50/4682]  lr: 4.954910e-06  eta: 5 days, 10:51:05  time: 0.479129  data_time: 0.216504  memory: 4547  loss: nan  loss_kpt: nan  acc_pose: 0.000000

**Comment by Limbo844** [comment]
补充：实验结果 在3090ti batch_size=64 学习率lr=5e-4 、 5e-5、5e-6、5e-7 下实验 loss均为nan
预计是JSDiscretLoss 损失函数有问题，测试 更换为KeypointMSELoss  loss显示正常

**Comment by Limbo844** [comment]
没人嘛？？？？？？ @openmmlab-bot @openmmlab-test @Ben-Louis

**Comment by Ben-Louis** [comment]
试下把 batch size 加大？这个 config 用的 batch size 是 64x8=512, 你用 32 的 batch size 训练会不太稳定

**Comment by Ben-Louis** [comment]
我刚刚试了下单卡 batch size 32 训练，没出现 nan 啊
![image](https://github.com/open-mmlab/mmpose/assets/26127467/b834cf11-94d7-44a6-b047-aff6695e162b)
请问您有修改代码吗？

**Comment by Limbo844** [comment]
代码 我未进行修改&nbsp; 下午我再重新跑一下 给您详细的反馈 感谢您的邮件



------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Peng ***@***.***&gt;; 
发送时间: 2023年7月9日(星期天) 上午10:49
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [open-mmlab/mmpose] [Bug] MMPOSE中configs/body_2d_keypoint/integral_regression/coco/ipr_res50_dsnt-8xb64-210e_coco-256x256.py运行结果为loss: nan  loss_kpt: nan  acc_pose: 0.0000 00 (Issue #2517)





 
我刚刚试了下单卡 batch size 32 训练，没出现 nan 啊
 
 请问您有修改代码吗？

**Comment by Limbo844** [comment]
测试环境：RTX 3060TI
测试代码路径：configs/body_2d_keypoint/integral_regression/coco/ipr_res50_dsnt-8xb64-210e_coco-256x256.py
测试代码修改部分：
train_dataloader 中 batch_size 修改为32
测试结果：
![image](https://github.com/open-mmlab/mmpose/assets/67669621/6e90399c-faf9-466e-97dc-d97113aa68a5)

**Comment by Limbo844** [comment]
测试环境：NVIDIA GeForce RTX 3060 Ti  
python：3.8.16 
PyTorch: 1.13.1+cu117
CUDA Runtime 11.7
TorchVision: 0.14.1+cpu
 OpenCV: 4.7.0
 MMEngine: 0.7.4
 
测试代码路径：configs/body_2d_keypoint/integral_regression/coco/ipr_res50_dsnt-8xb64-210e_coco-256x256.py

测试代码修改部分：
重新 寻找mmpose中最新的代码对本人的源代码进行替换 主要进行替换 内容为一下几部分
1、MultipleLossWrapper 2、 SmoothL1Loss 3、 JSDiscretLoss 
测试结果：
![image](https://github.com/open-mmlab/mmpose/assets/67669621/f5ff4a56-ea9d-449f-8458-e86b363372b8)

loss 仍然出现 nan 的情况 且acc 为0

**Comment by Limbo844** [comment]
为 验证 是否是本人 环境或者 ipr 与 dsnt_head 代码出现问题 故同等环境下验证：
mmpose下 另外积分回归方法文件：
测试代码路径：configs/body_2d_keypoint/integral_regression/coco/ipr_res50_8xb64-210e_coco-256x256.py
测试代码修改：train_dataloader 中 batch_size 修改为32
测试结果：
![image](https://github.com/open-mmlab/mmpose/assets/67669621/9466eb23-6606-4787-8ff4-d23fe2f3c04f)

代码结果正常，且 acc 不为0

**Comment by Ben-Louis** [comment]
我在 3090/2060 上用 batch size 32/12 测试过 DSNT 的 config, 都不会出现 nan。建议您检查下有没有改过源代码。您使用 tools/train.py 进行训练，使用的 mmpose 不一定与当前目录下 `mmpose/` 文件夹中的代码一致，可能是以前安装过的版本。建议训练前用 `pip install -e .` 安装一下或者在训练命令前加上 `PYTHONPATH=.`

**Comment by Limbo844** [comment]
好的 感谢您的耐心解答 明天 我会按照您的方法 重新测试一下 祝您周末愉快~

**Comment by hanfei66520** [comment]
这个最后是怎么解决的，我使用mmsegmentation也是出现了nan的问题跟你的情况很类似。

**Comment by hanfei66520** [comment]
![image](https://github.com/open-mmlab/mmpose/assets/110291222/15a9957d-260a-499f-baa1-dc0c79d031a7)

**Comment by hanfei66520** [comment]
![image](https://github.com/open-mmlab/mmpose/assets/110291222/e4219c26-edc7-4542-aff6-e8714ac49991)
这是我的一个结果，会训练中间出现。

**Comment by shuishiwojiade** [comment]
你好，请问如何解决的，我mmpose已经是最新版，使用的训练配置文件是mmpose提供的ipr_res50_dsnt-8xb64-210e_coco-256x256.py，训练的第2个iteration就出现了NAN。 @Ben-Louis
